File: src/modules/analytics.ts
import { HttpClient } from '../http-client';
import {
  AnalyticsDaily,
  AnalyticsDailyOptions,
  AnalyticsOverview,
  AnalyticsOverviewOptions,
  HttpMethod,
  MsGineValidationError,
} from '../types';

export class AnalyticsModule {
  constructor(private readonly http: HttpClient) {}

  /**
   * Get an aggregated overview of message activity across all channels.
   *
   * Requires READ_MESSAGES or FULL_ACCESS scope on the API key.
   *
   * @example
   * ```ts
   * const overview = await client.analytics.overview({
   *   from: '2026-01-01',
   *   to: '2026-04-06',
   * });
   * console.log(overview.totals.sent, overview.channels.sms?.delivered);
   * ```
   */
  async overview(options: AnalyticsOverviewOptions = {}): Promise<AnalyticsOverview> {
    return this.http.request<AnalyticsOverview>({
      method: HttpMethod.GET,
      path: '/developers/analytics/overview',
      queryParams: { from: options.from, to: options.to },
    });
  }

  /**
   * Get per-day message volume and cost for the last N days.
   *
   * @example
   * ```ts
   * const daily = await client.analytics.daily({ days: 7 });
   * for (const point of daily.data) {
   *   console.log(point.date, point.sms.sent);
   * }
   * ```
   */
  async daily(options: AnalyticsDailyOptions = {}): Promise<AnalyticsDaily> {
    const days = options.days ?? 30;
    if (days < 1 || days > 90)
      throw new MsGineValidationError('days must be between 1 and 90', 'days');

    return this.http.request<AnalyticsDaily>({
      method: HttpMethod.GET,
      path: '/developers/analytics/daily',
      queryParams: { days },
    });
  }
}

File: go/analytics.go
package msgine

import (
	"context"
)

// AnalyticsService provides methods for querying messaging statistics.
type AnalyticsService struct {
	t *transport
}

// Overview returns aggregated messaging stats across all channels.
//
// Both From and To are optional ISO 8601 date strings (e.g. "2026-01-01").
// If omitted the API returns stats for the last 30 days.
func (s *AnalyticsService) Overview(ctx context.Context, opts AnalyticsOverviewOptions) (*AnalyticsOverview, error) {
	q := queryValues("from", opts.From, "to", opts.To)
	var result AnalyticsOverview
	if err := s.t.do(ctx, "GET", "/developers/analytics/overview", q, nil, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// Daily returns per-day message volume for the last N days (1–90, default 30).
func (s *AnalyticsService) Daily(ctx context.Context, days int) (*AnalyticsDaily, error) {
	if days < 1 || days > 90 {
		return nil, &ValidationError{Field: "days", Message: "days must be between 1 and 90"}
	}
	q := queryValues("days", days)
	var result AnalyticsDaily
	if err := s.t.do(ctx, "GET", "/developers/analytics/daily", q, nil, &result); err != nil {
		return nil, err
	}
	return &result, nil
}
